arc/src/ArcToken.sol

```diff
@@ -0,0 +1,486 @@
+        address yieldToken;
+        // Set of all current token holders (for distribution purposes)
+        EnumerableSet.AddressSet holders;
+        // Added for asset name tracking
+        string assetName; // Name of the underlying asset (e.g., "Mineral Vault I")
+        // Token URI
+        string tokenURI;
+        // Purchase tracking
+        mapping(address => uint256) purchaseTimestamp; // When each holder purchased their tokens
+    }
+
+    // Calculate a unique storage slot for ArcTokenStorage (EIP-7201 standard).
+    // keccak256(abi.encode(uint256(keccak256("asset.token.storage")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant ARC_TOKEN_STORAGE_LOCATION =
+        0xf52c08b2e4132efdd78c079b339999bf65bd68aae758ed08b1bb84dc8f47c000;
+
+    function _getArcTokenStorage() private pure returns (ArcTokenStorage storage $) {
+        assembly {
+            $.slot := ARC_TOKEN_STORAGE_LOCATION
+        }
+    }
+
+    // -------------- Events --------------
+    event WhitelistStatusChanged(address indexed account, bool isWhitelisted);
+    event TransfersRestrictionToggled(bool transfersAllowed);
+    event YieldDistributed(uint256 amount);
+    event YieldTokenUpdated(address indexed newYieldToken);
+    event AssetNameUpdated(string newAssetName);
+    event TokenURIUpdated(string newTokenURI);
+    event TokenPurchased(address indexed buyer, uint256 amount, uint256 timestamp);
+
+    // -------------- Initializer --------------
+    /**
+     * @dev Initialize the token with name, symbol, and supply.
+     *      The deployer becomes the default admin. Transfers are unrestricted by default.
+     * @param name_ Token name (e.g., "aMNRL")
+     * @param symbol_ Token symbol (e.g., "aMNRL")
+     * @param initialSupply_ Initial token supply to mint to the admin
+     * @param yieldToken_ Address of the ERC20 token for yield distribution (e.g., USDC).
+     *                    Can be address(0) if setting later.
+     * @param initialTokenHolder_ Address that will receive the initial token supply
+     */
+    function initialize(
+        string memory name_,
+        string memory symbol_,
+        uint256 initialSupply_,
+        address yieldToken_,
+        address initialTokenHolder_
+    ) public initializer {
+        __ERC20_init(name_, symbol_);
+        __AccessControl_init();
+        __UUPSUpgradeable_init();
+        __ReentrancyGuard_init();
+
+        ArcTokenStorage storage $ = _getArcTokenStorage();
+
+        // Set up roles
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _grantRole(MANAGER_ROLE, msg.sender);
+        _grantRole(YIELD_MANAGER_ROLE, msg.sender);
+        _grantRole(YIELD_DISTRIBUTOR_ROLE, msg.sender);
+
+        // Set initial transfer restriction (true = unrestricted transfers)
+        $.transfersAllowed = true;
+
+        // Set initial yield token if provided
+        if (yieldToken_ != address(0)) {
```

Comment:
Let's require that `yieldToken_ != address(0)`. Either way we don't need this `if` statement